src/coreclr/src/vm/eventtrace.cpp

```diff
@@ -3498,75 +3498,87 @@ BOOL ETW::TypeSystemLog::AddTypeToGlobalCacheIfNotExists(TypeHandle th, BOOL * p
                 return fSucceeded;
             }
         }
+    }
+
+    // Step 1: go from LoaderModule to hash of types.
+    Module * pLoaderModule = th.GetLoaderModule();
+    _ASSERTE(pLoaderModule != NULL);
+    LoggedTypesFromModule * pLoggedTypesFromModule = nullptr;
 
-        // Step 1: go from LoaderModule to hash of types.
+    {
+        CrstHolder _crst(GetHashCrst());
+        
+        pLoggedTypesFromModule = s_pAllLoggedTypes->allLoggedTypesHash.Lookup(pLoaderModule);
+    }
 
-        Module * pLoaderModule = th.GetLoaderModule();
-        _ASSERTE(pLoaderModule != NULL);
-        LoggedTypesFromModule * pLoggedTypesFromModule = s_pAllLoggedTypes->allLoggedTypesHash.Lookup(pLoaderModule);
+    if (pLoggedTypesFromModule == NULL)
+    {
+        pLoggedTypesFromModule = new (nothrow) LoggedTypesFromModule(pLoaderModule);
         if (pLoggedTypesFromModule == NULL)
         {
-            pLoggedTypesFromModule = new (nothrow) LoggedTypesFromModule(pLoaderModule);
-            if (pLoggedTypesFromModule == NULL)
-            {
-                // out of memory.  Bail on ETW stuff
-                *pfCreatedNew = FALSE;
-                return fSucceeded;
-            }
-
-            fSucceeded = FALSE;
-            EX_TRY
-            {
-                s_pAllLoggedTypes->allLoggedTypesHash.Add(pLoggedTypesFromModule);
-                fSucceeded = TRUE;
-            }
-            EX_CATCH
-            {
-                fSucceeded = FALSE;
-            }
-            EX_END_CATCH(RethrowCorruptingExceptions);
-            if (!fSucceeded)
-            {
-                *pfCreatedNew = FALSE;
-                return fSucceeded;
-            }
+            // out of memory.  Bail on ETW stuff
+            *pfCreatedNew = FALSE;
+            return fSucceeded;
         }
+    }
 
-        // Step 2: From hash of types, see if our TypeHandle is there already
-        TypeLoggingInfo typeLoggingInfoPreexisting = pLoggedTypesFromModule->loggedTypesFromModuleHash.Lookup(th);
+    fSucceeded = FALSE;
+    EX_TRY
+    {
+        CrstHolder _crst(GetHashCrst());
+        s_pAllLoggedTypes->allLoggedTypesHash.Add(pLoggedTypesFromModule);
```

Comment:
Assuming that multiple threads can call this method at the same time then it is possible the order of operations will look like this:
Thread 1: s_pAllLoggedTypes->allLoggedTypesHash.Lookup(pLoaderModule) returns NULL
Thread 2: s_pAllLoggedTypes->allLoggedTypesHash.Lookup(pLoaderModule) returns NULL
Thread 1: s_pAllLoggedTypes->allLoggedTypesHash.Add(pLoggedTypesFromModule)
Thread 1: add types to thread1's pLoggedTypesFromModule
Thread 2: s_pAllLoggedTypes->allLoggedTypesHash.Add(pLoggedTypesFromModule)

At this point the types that thread 1 added have been lost because thread 2 replaced the module logged types with an empty instance when it called Add(). The typical solution is to write the code like this:

    Thingy* pThingy = null;
    {
        CrstHolder()
        pThingy = dictionary.Lookup(key);
    }
    if(pThingy == null) 
    {
        pThingy = CreateNewThingy();
        {
            CrstHolder();
            Thingy* pThingyRecheck = dictionary.Lookup(key);
            if(pThingyRecheck == null) // usual case, no other thread created it first
            {
                dictionary.Add(key, pThingy);
            }
            else // rare case, another thread already added it when we left the lock
            {
                FreeThingy(pThingy);
                pThingy = pThingyRecheck;
            }
        }
    }

A 2nd minor issue, the current code calls Add() every time, even if the initial lookup was successful. It works, but it is a little unexpected and does unneeded work. My snippet above only calls Add() if the initial lookup fails.